docs: langflow assistant feature (#12439)
langflow-ai/langflow: docs/docs/Flows/langflow-assistant.mdx

---
title: Build flows and components with Langflow Assistant
slug: /langflow-assistant
---

import Icon from "@site/src/components/icon";

> **Langflow Assistant** is an in-app co-pilot pane within the **Playground**.

Reviewer: I would use a different term (like pair programmer or virtual assistant) than co-pilot since that's Microsoft's pair programmer.

Like other co-pilots, **Langflow Assistant** can answer questions about the application and help you get more out of Langflow.

> **Langflow Assistant** understands the structure of the Langflow graph, so it can create flows and components from natural language prompts.

Reviewer: Optional, but users might want to know how this actually works. Is it a Langflow flow? What data can it read (can it see your environment variable values or only the env var names, can it access your whole LF deployment and other flows, etc)?

## Prerequisites

* Start Langflow with `LANGFLOW_AGENTIC_EXPERIENCE=true` set in your `.env` file.
* Connect an LLM provider in the [Global model providers page](/components-models).

## Create a custom component with Langflow Assistant

In this example, you'll prompt **Langflow Assistant** to create a custom component that validates and normalizes a list of URLs.

You'll then iterate on the code based on the results in the **Playground**.

1. In the **Playground**, click the <Icon name="Langflow" aria-hidden="true"/> **Langflow** icon.
    The **Langflow Assistant** pane opens.
2. Optionally, ask `What can you help me with?` for a list of **Langflow Assistant's** capabilities.
3. Prompt **Langflow Assistant** to generate a custom component. For example:
    ```
    Create a custom component URLTitleExtractor with:
    input: text
    output: list of {url, title, status}
    timeout handling + per-URL error handling
    clean docstring and typed methods.”
    ```

> 4. **Langflow Assistant** creates the component's code.

Reviewer: Unless the result is always the same, I would add some clarification that the user's actual result might vary from this example.

            return title
    ```

    </details>

    To inspect the code, click **View Code**.
    To add the component to the canvas, click **Add to Canvas**.

6. Connect the component to **Chat Input** and **Chat Output** components.
    At this point your flow has three connected components:

    * **Chat Input** sends the text that contains URLs into the **URLTitleExtractor** so the flow runs when you chat in the **Playground** or send a prompt from an app.
    * **URLTitleExtractor** reads that text, finds URLs, fetches each page, and outputs a [Table](/data-types#table) with `url`, `title`, and `status` columns.
    * **Chat Output** receives **URLTitleExtractor**'s result and displays the [Table](/data-types#table) back to the user or calling application.

    ![A custom URL Title Extractor component connected to a chat input and chat output](/img/lf-assistant-flow.png)

7. Open the **Playground**, and tell Langflow to check a list of URLs.
    For example:
    ```
    Check these links: https://langflow.org
    https://github.com/langflow-ai/langflow
    https://python.org
    https://this-domain-should-not-resolve-12345.invalid
    ```
8. Run the flow.

>     The output will be similar to the following, with a Table with one row per URL, including page titles and status codes.

Reviewer: ```suggestion
    The output will be similar to the following, with a table with one row for each URL, including page titles and status codes.
```

    ```text
    | url | title | status |
    | --- | --- | --- |
    | https://langflow.org | Langflow \| Low-code AI builder for agentic and RAG applications | 200 |
    | https://github.com/langflow-ai/langflow | GitHub - langflow-ai/langflow: Langflow is a powerful tool for building and deploying AI-powered agents and workflows. - GitHub | 200 |
    | https://python.org | Welcome to Python.org | 200 |
    | https://this-domain-should-not-resolve-12345.invalid |  |  |
    ```
9. To iterate further, tell **Langflow Assistant** what you want.
    For example, prompt it to `Update URLTitleExtractor to add max_urls (default 5) and skip duplicates.`.

>     **Langflow Assistant** creates a new **URLTitleExtractor** component, with the new features added.

Reviewer: ```suggestion
    **Langflow Assistant** creates a new version of the **URLTitleExtractor** component with the new features added.
```

> 10. Replace the old component with the new component.

Reviewer: ```suggestion

10. Replace the old component with the new component.
```

    Set **max_urls** to `3`.
    In the **Playground**, enter a list that includes a duplicate URL and the invalid URL from before:
    ```
    Check these links: https://langflow.org
    https://github.com/langflow-ai/langflow
    https://python.org
    https://langflow.org
    https://this-domain-should-not-resolve-12345.invalid
    ```

11. Run the flow.

>     The output will be similar to the following, with the duplicate `https://langflow.org` removed, and only 3 valid URLs displayed.

Reviewer: ```suggestion
    The output will be similar to the following, with the duplicate `https://langflow.org` removed and only 3 valid URLs displayed.
```

>     | url | title | status |

Reviewer: did you mean this to be in a code block?
